### PR TITLE
68 20230919 replication configuration

### PR DESCRIPTION
**Pull request by Firstyear**
Relates #68 - this adds the start of configuration and mtls for replication. I'll probably contain this to some early config and test setup to start with. 

Checklist

- [ x ] This pr contains no AI generated code
- [ x ] cargo fmt has been run
- [ ] cargo clippy has been run
- [ ] cargo test has been run and passes
- [ ] book chapter included (if relevant)
- [ ] design document included (if relevant)

### CODE REVIEW

File: server/core/src/config.rs [hunk 1]
```diff
@@ -179,41 +225,64 @@ pub struct Configuration {
     pub role: ServerRole,
     pub output_mode: ConsoleOutputMode,
     pub log_level: LogLevel,
+
+    /// Replication settings.
+    pub repl_config: Option<ReplicationConfiguration>,
+    /// This allows interally setting some unsafe options for replication.
+    pub integration_repl_config: Option<Box<IntegrationReplConfig>>,
```
Comment (yaleman) [inline server/core/src/config.rs:232]: We probably should be gating the existence of this behind `#[cfg(any(debug_assertions,test))]` so it's not triggerable in prod...

File: server/testkit/tests/mtls_test.rs [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+/// Gets an [EcGroup] for P-256
+pub fn get_group() -> Result<ec::EcGroup, OpenSSLError> {
+    ec::EcGroup::from_curve_name(Nid::X9_62_PRIME256V1)
+}
+
+fn keylog_cb(_ssl_ref: &SslRef, key: &str) {
+    trace!(?key);
+}
+
+fn build_self_signed_client_server(
```
Comment (yaleman) [inline server/testkit/tests/mtls_test.rs:41]: Perhaps `build_self_signed_client_cert` ?

File: server/core/src/config.rs (hunk 1 above)
Comment (Firstyear) [reply server/core/src/config.rs:232]: That's going to be a more fun change I expect .... 🤔 but it's possible.

File: server/testkit/tests/mtls_test.rs (hunk 2 above)
Comment (Firstyear) [reply server/testkit/tests/mtls_test.rs:41]: It's going to move shortly I think as I need it in the main daemon for building the certs inside Kanidm, and allowing them to be exposed to other partso fthe code.

Also it's a hybrid client and server cert? Maybe a different name needed?

Comment (yaleman) [reply server/testkit/tests/mtls_test.rs:41]: Yeah, that's more the point - it wasn't exactly clear without reading the code what it was doing, and later it's called and I thought we were starting up a server with a self-signed cert or something 😄

**Comment by Firstyear** [comment]
@yaleman running the two nodes on localhost with:

```
CONFIG_FILE=/Users/william/tmp/insecure_server_secondary.toml  ./run_insecure_dev_server.sh
CONFIG_FILE=/Users/william/tmp/insecure_server_primary.toml ./run_insecure_dev_server.sh
```

```
# cat /Users/william/tmp/insecure_server_primary.toml
bindaddress = "127.0.0.1:8443"
ldapbindaddress = "127.0.0.1:3636"

db_fs_type = "zfs"
db_path = "/Users/william/tmp/kanidm-primary.db"
tls_chain = "/tmp/kanidm/chain.pem"
tls_key = "/tmp/kanidm/key.pem"

#   The log level of the server. May be one of info, debug, trace
#
#   NOTE: this is overridden by environment variables at runtime
#   Defaults to "info"
#
# log_level = "info"
log_level = "debug"
# log_level = "trace"

domain = "localhost"
origin = "https://localhost:8443"
trust_x_forward_for = true

i_acknowledge_that_replication_is_in_development = true

[replication]
origin = "https://localhost:8444"
bindaddress = "127.0.0.1:8444"

[replication."https://localhost:9444"]
type = "allow-pull"
consumer_cert = "<fill this in with the content ofCONFIG_FILE=/Users/william/tmp/insecure_server_secondary.toml ./run_insecure_dev_server.sh  show-replication-certificate> "

[online_backup]
path = "/tmp/kanidm/backups/"
schedule = "@hourly"
```

```
# cat /Users/william/tmp/insecure_server_secondary.toml
bindaddress = "127.0.0.1:9443"
ldapbindaddress = "127.0.0.1:3936"

db_fs_type = "zfs"
db_path = "/Users/william/tmp/kanidm-secondary.db"
tls_chain = "/tmp/kanidm/chain.pem"
tls_key = "/tmp/kanidm/key.pem"

#   The log level of the server. May be one of info, debug, trace
#
#   NOTE: this is overridden by environment variables at runtime
#   Defaults to "info"
#
# log_level = "info"
log_level = "debug"
# log_level = "trace"

domain = "localhost"
origin = "https://localhost:9443"
trust_x_forward_for = true

i_acknowledge_that_replication_is_in_development = true

[replication]
origin = "https://localhost:9444"
bindaddress = "127.0.0.1:9444"

[replication."https://localhost:8444"]
type = "pull"
supplier_cert = "<fill this in with the content of CONFIG_FILE=/Users/william/tmp/insecure_server_primary.toml ./run_insecure_dev_server.sh  show-replication-certificate>"
automatic_refresh = true

[online_backup]
path = "/tmp/kanidm/backups/"
schedule = "@hourly"
```

**Comment by yaleman** [comment]
I can't build this... there's some files missing?

```
error[E0583]: file not found for module `mtls`
  --> libs/crypto/src/lib.rs:32:1
   |
32 | pub mod mtls;
   | ^^^^^^^^^^^^^
   |
   = help: to create the module `mtls`, create file "libs/crypto/src/mtls.rs" or "libs/crypto/src/mtls/mod.rs"

error[E0583]: file not found for module `prelude`
  --> libs/crypto/src/lib.rs:33:1
   |
33 | pub mod prelude;
   | ^^^^^^^^^^^^^^^^
   |
   = help: to create the module `prelude`, create file "libs/crypto/src/prelude.rs" or "libs/crypto/src/prelude/mod.rs"

error[E0583]: file not found for module `serialise`
  --> libs/crypto/src/lib.rs:34:1
   |
34 | pub mod serialise;
   | ^^^^^^^^^^^^^^^^^^
   |
   = help: to create the module `serialise`, create file "libs/crypto/src/serialise.rs" or "libs/crypto/src/serialise/mod.rs"
```

**Comment by yaleman** [comment]
Also keystorage

**Comment by Firstyear** [comment]
Okay, there are only three more things for replication:
* Make the RUV actually persistent to the database
* Fix up merge_state in entry to handle valuesets that have merge support.
* Allow admins to perform a manual refresh from the admin cli

**Comment by Firstyear** [comment]
Oh and lots of tidy up around timeouts, cert gen and how long they are for, automatic renewals and that kind of thing.

File: server/daemon/src/main.rs [hunk 3]
```diff
@@ -145,6 +147,22 @@ async fn submit_admin_req(path: &str, req: AdminTaskRequest, output_mode: Consol
...
         },
+        Some(Ok(AdminTaskResponse::ShowReplicationCertificate { cert })) => match output_mode {
+            ConsoleOutputMode::JSON => {
+                eprintln!("{{\"certificate\":\"{}\"}}", cert)
+            }
+            ConsoleOutputMode::Text => {
+                info!(certificate = ?cert)
+            }
+        },
+        Some(Ok(AdminTaskResponse::RenewReplicationCertificate { success })) => match output_mode {
```
Comment (yaleman) [inline server/daemon/src/main.rs:158]: We really should show the cert here 'cause folks are going to be pulling it next anyhow.

File: server/daemon/src/main.rs [hunk 4]
```diff
@@ -531,6 +561,26 @@ async fn main() -> ExitCode {
...
+                    let output_mode: ConsoleOutputMode = commonopts.output_mode.to_owned().into();
+                    submit_admin_req(config.adminbindpath.as_str(),
+                        AdminTaskRequest::ShowReplicationCertificate,
+                        output_mode,
+                    ).await;
+                }
+                KanidmdOpt::RenewReplicationCertificate {
+                    commonopts
+                } => {
+                    info!("Running show replication certificate ...");
```
Comment (yaleman) [inline server/daemon/src/main.rs:577]:
```suggestion
                    info!("Running renew replication certificate ...");
```

File: server/daemon/src/opt.rs [hunk 5]
```diff
@@ -154,6 +154,16 @@ enum KanidmdOpt {
         #[clap(flatten)]
         commonopts: CommonOpt,
     },
+    /// Display this servers replication certificate
```
Comment (yaleman) [inline server/daemon/src/opt.rs:157]:
```suggestion
    /// Display this server's replication certificate
```

File: server/daemon/src/opt.rs [hunk 6]
```diff
@@ -154,6 +154,16 @@ enum KanidmdOpt {
...
+    ShowReplicationCertificate {
+        #[clap(flatten)]
+        commonopts: CommonOpt,
+    },
+    /// Renew this servers replication certificate
```
Comment (yaleman) [inline server/daemon/src/opt.rs:162]:
```suggestion
    /// Renew this server's replication certificate
```

File: server/lib/src/be/mod.rs [hunk 7]
```diff
@@ -872,10 +873,13 @@ pub trait BackendTransaction {
...
             .and_then(|u| u.ok_or(OperationError::InvalidDbState))?;
 
-        let bak = DbBackup::V2 {
+        let keyhandles = idlayer.get_key_handles()?;
+
+        let bak = DbBackup::V3 {
             db_s_uuid,
             db_d_uuid,
             db_ts_max,
+            keyhandles,
```
Comment (yaleman) [inline server/lib/src/be/mod.rs:882]: Can we call it something more ... clear? I know *you* know what it means and I know what it means in this very specific context, but ... `keyhandles` ... is confusing.

Maybe `replication_keys`? There's a lot of keys in this system, and I hope we're not stuffing them all in one place 😄

File: server/testkit/tests/mtls_test.rs [hunk 8]
```diff
@@ -0,0 +1,277 @@
...
+async fn test_mtls_client_without_server_ca() {
+    let (mut tlsstream, handle, _tx) = setup_mtls_test(TestCase::ClientWithoutServerCa).await;
+
+    let err: openssl::ssl::Error = SslStream::connect(Pin::new(&mut tlsstream))
+        .await
+        .unwrap_err();
+    trace!(?err);
+    // Tls Post Process Certificate (Certificate Verify Failed)
+    let ossl_err = err.ssl_error().and_then(|e| e.errors().get(0)).unwrap();
+    assert!(ossl_err.code() == 2147483650);
```
Comment (yaleman) [inline server/testkit/tests/mtls_test.rs:268]: This test is unreliable as the error codes depend on library versions etc (both my machine and CI threw different error codes)
```suggestion
    assert!(ossl_err.reason() == Some("certificate verify failed"));
```

Comment (Firstyear) [reply server/testkit/tests/mtls_test.rs:268]: :O 
I am shoock. I tried to do the code to be *extremely* specific about the error, but it's wild that the code changed between different versions of the lib.

File: server/lib/src/be/mod.rs (hunk 7 above)
Comment (Firstyear) [reply server/lib/src/be/mod.rs:882]: The reason it's keyhandles is that in the future I plan to use this for persistent storage of keyhandles that are loaded into a tpm. So keyhandles is meant to be "here is where we persist key material". 

Which btw, is getting a bit into how a TPM works, but they don't store anything internally. You have to temporarily load them in and out, and we have to store the encrypted blobs ourselves. 

Anyway, that's why it's keyhandles.

File: server/daemon/src/main.rs (hunk 3 above)
Comment (Firstyear) [reply server/daemon/src/main.rs:158]: Yep, that's doable :)

File: server/testkit/tests/mtls_test.rs (hunk 8 above)
Comment (yaleman) [reply server/testkit/tests/mtls_test.rs:268]: Yeah I'm digging into how/why it changes, but it's definitely a thing.

File: server/lib/src/be/mod.rs (hunk 7 above)
Comment (yaleman) [reply server/lib/src/be/mod.rs:882]: Ok, cool, thx 😄

**Review by yaleman — APPROVED** [review]
LGTM

**Pull request merged**